src/roots.rs

```diff
@@ -25,41 +25,91 @@ impl Roots {
         path.push(name);
 
         debug!("Adding root from {:?} to {:?}", store_path, path,);
-        ignore_missing(std::fs::remove_file(&path))?;
-        symlink(&store_path, &path)?;
+        check_permission(
+            ignore_missing(std::fs::remove_file(&path)),
+            &format!("Can’t remove {}", path.display()),
+        )?;
+
+        check_permission(
+            symlink(&store_path, &path),
+            &format!(
+                "Can’t symlink {} to {}",
+                store_path.display(),
+                path.display()
+            ),
+        )?;
 
         // this is bad.
         let mut root = PathBuf::from("/nix/var/nix/gcroots/per-user");
+        // TODO: check on start
         root.push(env::var("USER").expect("env var 'USER' must be set"));
+
+        // The user directory sometimes doesn’t exist,
+        // but we can create it (it’s root but `rwxrwxrwx`)
+        ensure_directory_exists(&root);
+
         root.push(format!("{}-{}", self.id, name));
 
         debug!("Connecting root from {:?} to {:?}", path, root,);
-        ignore_missing(std::fs::remove_file(&root))?;
-        symlink(&path, &root)?;
+        check_permission(
```

Comment:
I don't understand why we panic on permission denied, where currently we otherwise pass an `io` error on up anyway. What is the purpose of adding the extra handling of the EPERM?

Reply:
The problem is that the panic errors by `io` are extremely nondescriptive, so I’d rather fail fast and tell the user the reason. We have to fail anyway, and this is I/O code.

I mean we could split them out into their own enum and handle them on the top layer, but I don’t see anything we gain from that at this stage (can be done later if needed).

Reply:
Actually, let me rebase on #55 and push it via the new interface.

Reply:
You're right that IO errors are quite nondescript. I wrote up an example of a nicer, more rust-like way of implementing this: https://github.com/target/lorri/commit/04121896b6bebb44bdb06350ff792da85f57d8c5 Check it out?